common/src/main/java/com/microsoft/identity/common/internal/platform/DevicePopManager.java

```diff
@@ -0,0 +1,834 @@
+                TAG,
+                clientException.getMessage(),
+                clientException
+        );
+
+        throw clientException;
+    }
+
+    //region Internal Functions
+
+    /**
+     * Generates a new RSA KeyPair of the specified lenth.
+     *
+     * @param ctx        The current application Context.
+     * @param minKeySize The minimum keysize to use.
+     * @return The newly generated RSA KeyPair.
+     * @throws UnsupportedOperationException
+     */
+    @SuppressLint("NewApi")
+    private KeyPair generateNewRsaKeyPair(@NonNull final Context ctx,
+                                          final int minKeySize)
+            throws UnsupportedOperationException, InvalidAlgorithmParameterException,
+            NoSuchAlgorithmException, NoSuchProviderException {
+        final int MAX_RETRIES = 4;
+
+        for (int ii = 0; ii < MAX_RETRIES; ii++) {
+            KeyPair kp;
+
+            try {
+                kp = generateNewKeyPair(ctx, true);
+            } catch (final StrongBoxUnavailableException e) {
+                Logger.error(
+                        TAG,
+                        "StrongBox unsupported - skipping hardware flags.",
+                        e
+                );
+
+                // Retry, but don't request StrongBox
+                kp = generateNewKeyPair(ctx, false);
+            }
+
+            // Due to a bug in some versions of Android, keySizes may not be exactly as specified
+            // To generate a 2048-bit key, two primes of length 1024 are multiplied -- this product
+            // may be 2047 in length in some cases which causes Nimbus to crash. To avoid this,
+            // check the keysize prior to returning the generated KeyPair.
+
+            // Since this seems to be nondeterministic in nature, attempt this a maximum of 4 times.
+            final int length = RSAKeyUtils.keyBitLength(kp.getPrivate());
+
+            // If the key material is hidden (HSM or otherwise) the length is -1
+            if (length >= minKeySize || length < 0) {
+                logSecureHardwareState(kp);
+
+                return kp;
+            }
+        }
+
+        // Clean up... we generated a cert, but it cannot be used.
+        clearAsymmetricKey();
+
+        throw new UnsupportedOperationException(
+                "Failed to generate valid KeyPair. Attempted " + MAX_RETRIES + " times."
+        );
+    }
+
+    private void logSecureHardwareState(@NonNull final KeyPair kp) {
+        String msg;
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            try {
+                final PrivateKey privateKey = kp.getPrivate();
+                final KeyFactory factory = KeyFactory.getInstance(
+                        privateKey.getAlgorithm(), ANDROID_KEYSTORE
+                );
+                final KeyInfo info = factory.getKeySpec(privateKey, KeyInfo.class);
+                final boolean isInsideSecureHardware = info.isInsideSecureHardware();
```

Comment:
can we use this information to decide if we should try generate a key with useStrongBox=true?